src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +148,331 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
   }
   return entry;
 }
+
+TemplateAssertionPredicateBool::TemplateAssertionPredicateBool(Node* source_bool) : _source_bool(source_bool->as_Bool()) {
+#ifdef ASSERT
+  // During IGVN, we could have multiple outputs of the _source_bool, for example, when the backedge of the loop of
+  // this Template Assertion Predicate is about to die and the CastII on the last value bool already folded to a
+  // constant (i.e. no OpaqueLoop* nodes anymore). Then IGVN could already have commoned up the bool with the bool of
+  // one of the Hoisted Check Predicates. Just check that the Template Assertion Predicate is one of the outputs.
+  bool has_template_output = false;
+  for (DUIterator_Fast imax, i = source_bool->fast_outs(imax); i < imax; i++) {
+    Node* out = source_bool->fast_out(i);
+    if (out->Opcode() == Op_Opaque4) {
+      has_template_output = true;
+      break;
+    }
+  }
+  assert(has_template_output, "must find Template Assertion Predicate as output");
```

Comment:
Idea: pack this in a method `has_opaque4_output`. Then you do not need the `has_template_output` variable, but can directly `return true` and `return false` at the end. And call like that:
`DEBUG_ONLY(source_bool->has_opaque4_output();)`
The nice thing: the constructor could be moved to the `hpp` file.